lib/TokenManager.ts

```diff
@@ -287,9 +287,28 @@ export class TokenManager {
     this.off = tokenMgmtRef.emitter.off.bind(tokenMgmtRef.emitter);
     this.hasExpired = hasExpired.bind(this, tokenMgmtRef);
   
+    let tooManyRenews = false;
+    const renewTimeQueue = [];
     const onTokenExpiredHandler = (key) => {
       if (options.autoRenew) {
-        this.renew(key).catch(() => {}); // Renew errors will emit an "error" event 
+        // detect if there are too many renew requests
+        if (renewTimeQueue.length < 10) {
+          renewTimeQueue.push(+new Date());
+        } else {
+          // get and remove first item from queue
+          const firstTime = renewTimeQueue.shift();
+          const lastTime = renewTimeQueue[renewTimeQueue.length - 1];
+          if (lastTime - firstTime < 30 * 1000) {
+            tooManyRenews = true;
+          }
+        }
+        
+        if (tooManyRenews) {
```

Comment:
So if there are 10+ in the queue, we will absolutely remove the first, but we may not decide there are too many.  (and we aren't putting "now" into the queue for that case)

Reply:
So here it's maintaining a queue with the max length of 10. When items in q reach 10, it compares if these 10 requests happen in a too short time window.
The flag here is just to prevent running the renew method again. When the next expired event comes, it just emit the error again, I think that what Aaron is concerned.

Reply:
But the queue is getting shorter (because of the shift), so the next time through tooManyRenews will be false (because it never gets evaluated, because there were only 9 items on the queue)

so in an infinite loop situation it would be:

1. Queue up first 10 items
2. 11th item emits error, but queue is reduced to 9
3. Queue up the 12 item as the 10th in the queue
3. 13 item emits error, and queue is again reduced to 9

Reply:
the tooManyRenews is initialized in the constructor, which only be assigned once no matter how many expired events emitted. In the current implementation, the value won't be reassigned after set to true. I was initially thought that would be enough to prevent the infinite renew calls, but it does introduce side effect like keep emitting errors.